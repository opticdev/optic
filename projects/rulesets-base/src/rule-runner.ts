import { IFact, IChange, OpenAPIV3, Result } from '@useoptic/openapi-utilities';
import {
  Ruleset,
  SpecificationRule,
  OperationRule,
  RequestRule,
  ResponseRule,
} from './rules';

// TODO maybe create a class instead
export const createRuleRunner = (
  rules: (
    | Ruleset
    | SpecificationRule
    | OperationRule
    | RequestRule
    | ResponseRule
  )[]
) => ({
  // Matches the legacy rule runner interface
  runRulesWithFacts: (options: {
    context: any;
    nextFacts: IFact[];
    currentFacts: IFact[];
    changelog: IChange[];
    nextJsonLike: OpenAPIV3.Document;
    currentJsonLike: OpenAPIV3.Document;
  }): Result[] => {
    // TODO implement rule runner
    return [];
  },
});
