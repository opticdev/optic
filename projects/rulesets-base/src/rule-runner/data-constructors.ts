/**
 * data-constructors take a Node from OpenAPIFactNodes and a before / after spec
 * to construct it into a user facing data structure used in `matches` blocks and `rule assertions`
 * some examples of these user facing data structures are `Operation`, `Specification`, `Response`
 */
import {
  OpenApiKind,
  OpenAPIV3,
  FactVariant,
} from '@useoptic/openapi-utilities';
import { jsonPointerHelpers } from '@useoptic/json-pointer-helpers';
import {
  Operation,
  Specification,
  RequestBody,
  Response,
  ResponseBody,
} from '../types';
import {
  OpenAPIFactNodes,
  EndpointNode,
  RequestNode,
  ResponseNode,
  NodeDetail,
  BodyNode,
} from './rule-runner-types';

type SpecFactsFrom = 'before' | 'after';

const createFactsWithRaw = <T extends OpenApiKind>(
  nodeDetailMap: Map<string, NodeDetail<T>>,
  key: SpecFactsFrom,
  openApiSpec: OpenAPIV3.Document
): Map<string, FactVariant<T> & { raw: any }> => {
  const factsWithRaw: Map<string, FactVariant<T> & { raw: any }> = new Map();

  for (const [name, nodeDetail] of nodeDetailMap.entries()) {
    const nodeFact = nodeDetail[key];
    if (nodeFact) {
      factsWithRaw.set(name, {
        ...nodeFact,
        raw: jsonPointerHelpers.get(openApiSpec, nodeFact.location.jsonPath),
      });
    }
  }

  return factsWithRaw;
};
const createPropertyFactsWithRaw = <T extends OpenApiKind>(
  nodeDetailMap: Map<string, NodeDetail<T>>,
  key: SpecFactsFrom,
  openApiSpec: OpenAPIV3.Document
): Map<
  string,
  FactVariant<T> & { raw: any; parent: OpenAPIV3.NonArraySchemaObject }
> => {
  const factsWithRaw: Map<
    string,
    FactVariant<T> & { raw: any; parent: OpenAPIV3.NonArraySchemaObject }
  > = new Map();

  for (const [name, nodeDetail] of nodeDetailMap.entries()) {
    const nodeFact = nodeDetail[key];
    if (nodeFact) {
      const parentPath = jsonPointerHelpers.pop(
        jsonPointerHelpers.pop(nodeFact.location.jsonPath)
      );
      factsWithRaw.set(name, {
        ...nodeFact,
        raw: jsonPointerHelpers.get(openApiSpec, nodeFact.location.jsonPath),
        parent: jsonPointerHelpers.get(openApiSpec, parentPath),
      });
    }
  }

  return factsWithRaw;
};

export const createRequest = (
  request: RequestNode,
  contentType: string,
  key: SpecFactsFrom,
  openApiSpec: OpenAPIV3.Document
): RequestBody | null => {
  const requestFact = request[key];
  const requestBody = request.bodies.get(contentType);
  const requestBodyFact = requestBody?.[key];
  if (!requestFact || !requestBody || !requestBodyFact) {
    return null;
  }
  const requestPointer = jsonPointerHelpers.compile([
    ...jsonPointerHelpers.decode(requestBodyFact.location.jsonPath).slice(0, 3),
    'requestBody',
  ]);
  const required: boolean | undefined = jsonPointerHelpers.get(
    openApiSpec,
    requestPointer
  ).required;

  return {
    ...requestBodyFact,
    required: required,
    raw: jsonPointerHelpers.get(openApiSpec, requestBodyFact.location.jsonPath),
    contentType,
    schemas: createPropertyFactsWithRaw(requestBody.schemas, key, openApiSpec),
    properties: createPropertyFactsWithRaw(
      requestBody.fields,
      key,
      openApiSpec
    ),
  };
};

export const createResponse = (
  response: ResponseNode,
  key: SpecFactsFrom,
  openApiSpec: OpenAPIV3.Document
): Response | null => {
  const responseFact = response[key];
  if (!responseFact) {
    return null;
  }

  const bodies: ResponseBody[] = [];
  for (const [contentType, bodyNode] of response.bodies.entries()) {
    const responseBody = createResponseBody(
      bodyNode,
      response.statusCode,
      contentType,
      key,
      openApiSpec
    );

    if (responseBody) {
      bodies.push(responseBody);
    }
  }

  return {
    ...responseFact,
    raw: jsonPointerHelpers.get(openApiSpec, responseFact.location.jsonPath),
    statusCode: response.statusCode,
    bodies,
    headers: createFactsWithRaw(response.headers, key, openApiSpec),
  };
};

export const createResponseBody = (
  bodyNode: BodyNode,
  statusCode: string,
  contentType: string,
  key: SpecFactsFrom,
  openApiSpec: OpenAPIV3.Document
): ResponseBody | null => {
  const responseBodyFact = bodyNode?.[key];
  if (!responseBodyFact) {
    return null;
  }

  return {
    ...responseBodyFact,
    raw: jsonPointerHelpers.get(
      openApiSpec,
      responseBodyFact.location.jsonPath
    ),
    contentType,
    statusCode: statusCode,
    properties: createPropertyFactsWithRaw(bodyNode.fields, key, openApiSpec),
    schemas: createPropertyFactsWithRaw(bodyNode.schemas, key, openApiSpec),
  };
};

export const createOperation = (
  endpoint: EndpointNode,
  key: SpecFactsFrom,
  openApiSpec: OpenAPIV3.Document
): Operation | null => {
  const operationFact = endpoint[key];

  if (!operationFact) {
    return null;
  }

  const requests: RequestBody[] = [];
  const responses = new Map<string, Response>();

  for (const contentType of endpoint.request.bodies.keys()) {
    const request = createRequest(
      endpoint.request,
      contentType,
      key,
      openApiSpec
    );
    if (request) {
      requests.push(request);
    }
  }

  for (const responseNode of endpoint.responses.values()) {
    const response = createResponse(
      responseNode,

      key,
      openApiSpec
    );
    if (response) {
      responses.set(responseNode.statusCode, response);
    }
  }

  const security: OpenAPIV3.OperationObject['security'] | null = (() => {
    const operationOverride = jsonPointerHelpers.tryGet(
      openApiSpec,
      jsonPointerHelpers.append(operationFact.location.jsonPath, 'security')
    );

    if (operationOverride.match)
      return operationOverride.value as OpenAPIV3.OperationObject['security'];

    const specLevel = jsonPointerHelpers.tryGet(openApiSpec, '/security');

    if (specLevel.match)
      return specLevel.value as OpenAPIV3.OperationObject['security'];

    return null;
  })();

  return {
    ...operationFact,
    polymorphicSchemas: endpoint.polymorphicSchemas,
    raw: jsonPointerHelpers.get(openApiSpec, operationFact.location.jsonPath),
    path: endpoint.path,
    method: endpoint.method,
    security,
    headerParameters: createFactsWithRaw(
      endpoint.headerParameters,
      key,
      openApiSpec
    ),
    queryParameters: createFactsWithRaw(
      endpoint.queryParameters,
      key,
      openApiSpec
    ),
    pathParameters: createFactsWithRaw(
      endpoint.pathParameters,
      key,
      openApiSpec
    ),
    cookieParameters: createFactsWithRaw(
      endpoint.cookieParameters,
      key,
      openApiSpec
    ),
    requests,
    responses,
  };
};

export const createSpecification = (
  specificationNode: OpenAPIFactNodes['specification'],
  key: SpecFactsFrom,
  openApiSpec: OpenAPIV3.Document
): Specification | null => {
  const specificationFact = specificationNode[key];
  if (!specificationFact) {
    return null;
  }

  return {
    ...specificationFact,
    raw: openApiSpec,
  };
};
