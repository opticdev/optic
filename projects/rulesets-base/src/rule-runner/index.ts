import { Matchers, Matcher } from './matchers/utils';
export { Matchers, Matcher };
export * from './rule-runner';
