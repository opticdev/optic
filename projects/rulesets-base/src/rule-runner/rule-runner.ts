import {
  IFact,
  IChange,
  Result,
  OperationLocation,
  ILocation,
  OpenApiKind,
  ObjectDiff,
  RuleResult,
  traverseSpec,
  factsToChangelog,
  Severity,
} from '@useoptic/openapi-utilities';
import {
  ISpectralDiagnostic,
  Spectral,
  RulesetDefinition as SpectralRulesetDefinition,
} from '@stoplight/spectral-core';
import { oas } from '@stoplight/spectral-rulesets';

import { jsonPointerHelpers } from '@useoptic/json-pointer-helpers';
import isEqual from 'lodash.isequal';

import { groupFacts } from './group-facts';
import { runSpecificationRules } from './specification';
import { runOperationRules } from './operation';
import { runRequestRules } from './request';
import { runResponseBodyRules } from './response-body';
import { runResponseRules } from './response';
import { ExternalRule, Rule, Ruleset } from '../rules';
import { ExternalRuleBase } from '../rules/external-rule-base';
import { OpenAPIDocument } from '..';

type SpectralRules = Extract<
  SpectralRulesetDefinition,
  { extends: any; rules: any }
>['rules'];

function resultToRuleResult(r: Result): RuleResult {
  return {
    where: r.where,
    error: r.error,
    passed: r.passed,
    exempted: r.exempted,
    location: {
      jsonPath: r.change.location.jsonPath,
      spec:
        'removed' in r.change && !!r.change.removed?.before
          ? 'before'
          : 'after',
    },
    name: r.name ?? 'Rule',
    type: r.type,
    docsLink: r.docsLink,
    expected: r.expected,
    received: r.received,
    severity: r.severity ?? Severity.Error,
  };
}

export class RuleRunner {
  constructor(private rules: (Ruleset | Rule | ExternalRule)[]) {}

  // TODO deprecate this once spectral rules are natively supported
  async runSpectralRules({
    ruleset,
    nextFacts,
    nextJsonLike,
  }: {
    ruleset: SpectralRules;
    nextFacts: IFact[];
    nextJsonLike: OpenAPIDocument;
  }): Promise<Result[]> {
    if ((nextJsonLike as any)['x-optic-ci-empty-spec'] === true) {
      return [];
    }
    const spectral = new Spectral();
    spectral.setRuleset({
      extends: [[oas as SpectralRulesetDefinition, 'all']],
      rules: ruleset,
    });
    const operations = nextFacts.filter(
      (i) => i.location.kind === OpenApiKind.Operation
    );

    const results: ISpectralDiagnostic[] = await spectral.run(
      nextJsonLike as any
    );
    const opticResult: Result[] = results.map((spectralResult) => {
      const operationPath = spectralResult.path.slice(0, 3);
      const matchingOperation = operations.find((i) =>
        isEqual(i.location.conceptualPath, operationPath)
      );

      const location: ILocation = {
        conceptualLocation: (matchingOperation
          ? matchingOperation.location.conceptualLocation
          : { path: 'This Specification', method: '' }) as OperationLocation,
        jsonPath: jsonPointerHelpers.compile(
          spectralResult.path.map((i) => i.toString())
        ),
        conceptualPath: [],
        kind: 'API' as any,
      } as any;

      return {
        condition: spectralResult.code.toString(),
        passed: false,
        error: spectralResult.message,
        isMust: true,
        isShould: false,
        where: 'requirement ',
        change: {
          location,
        } as any,
      };
    });

    return opticResult;
  }

  // TODO deprecate
  async runRulesWithFacts(inputs: {
    context: any;
    nextFacts: IFact[];
    currentFacts: IFact[];
    changelog: IChange[];
    nextJsonLike: OpenAPIDocument;
    currentJsonLike: OpenAPIDocument;
  }): Promise<Result[]> {
    const {
      context,
      currentFacts,
      nextFacts,
      changelog,
      currentJsonLike: beforeApiSpec,
      nextJsonLike: afterApiSpec,
    } = inputs;
    const externalRules = this.rules.filter((rule) =>
      ExternalRuleBase.isInstance(rule)
    ) as ExternalRule[];
    const rulesOrRulesets = this.rules.filter(
      (rule) => !ExternalRuleBase.isInstance(rule)
    ) as (Ruleset | Rule)[];

    // Groups the flat list of beforefacts, afterfacts and changes by location (e.g. operation, query parameter, response, response property, etc).
    // A node can contain a before fact, after fact and or change.
    const openApiFactNodes = groupFacts({
      beforeFacts: currentFacts,
      afterFacts: nextFacts,
      changes: changelog,
    });

    const externalResults: Result[] = [];
    for (const externalRule of externalRules) {
      const results = await externalRule.runRules({
        ...inputs,
        groupedFacts: openApiFactNodes,
      });
      externalResults.push(...results);
    }

    // Run rules on specifications and collect the results
    const specificationResults = runSpecificationRules({
      specificationNode: openApiFactNodes.specification,
      rules: rulesOrRulesets,
      customRuleContext: context,
      beforeApiSpec,
      afterApiSpec,
    });

    const endpointResults: Result[] = [];

    // For each endpoint from the endpoint fact nodes (this will include endpoints in both before and after specs) run rules and collect the results
    for (const endpointNode of openApiFactNodes.endpoints.values()) {
      const operationResults = runOperationRules({
        specificationNode: openApiFactNodes.specification,
        operationNode: endpointNode,
        rules: rulesOrRulesets,
        customRuleContext: context,
        beforeApiSpec,
        afterApiSpec,
      });
      endpointResults.push(...operationResults);

      const requestRules = runRequestRules({
        specificationNode: openApiFactNodes.specification,
        operationNode: endpointNode,
        requestNode: endpointNode.request,
        rules: rulesOrRulesets,
        customRuleContext: context,
        beforeApiSpec,
        afterApiSpec,
      });
      endpointResults.push(...requestRules);

      for (const responseNode of endpointNode.responses.values()) {
        const responseRules = runResponseRules({
          specificationNode: openApiFactNodes.specification,
          operationNode: endpointNode,
          responseNode: responseNode,
          rules: rulesOrRulesets,
          customRuleContext: context,
          beforeApiSpec,
          afterApiSpec,
        });

        const responseBodyRules = runResponseBodyRules({
          specificationNode: openApiFactNodes.specification,
          operationNode: endpointNode,
          responseNode: responseNode,
          rules: rulesOrRulesets,
          customRuleContext: context,
          beforeApiSpec,
          afterApiSpec,
        });
        endpointResults.push(...responseBodyRules, ...responseRules);
      }
    }

    return [...externalResults, ...specificationResults, ...endpointResults];
  }

  async runRules(inputs: {
    context: any;
    diffs: ObjectDiff[];
    fromSpec: OpenAPIDocument;
    toSpec: OpenAPIDocument;
  }): Promise<RuleResult[]> {
    const { context, fromSpec: beforeApiSpec, toSpec: afterApiSpec } = inputs;
    const externalRules = this.rules.filter((rule) =>
      ExternalRuleBase.isInstance(rule)
    ) as ExternalRule[];
    const rulesOrRulesets = this.rules.filter(
      (rule) => !ExternalRuleBase.isInstance(rule)
    ) as (Ruleset | Rule)[];

    // TODO reimplement the rule runner so we don't need to generate legacy fact types here
    const beforeFacts =
      inputs.fromSpec['x-optic-ci-empty-spec'] === true
        ? []
        : traverseSpec(inputs.fromSpec as any);
    const afterFacts =
      inputs.toSpec['x-optic-ci-empty-spec'] === true
        ? []
        : traverseSpec(inputs.toSpec as any);
    const changelog = factsToChangelog(beforeFacts, afterFacts);

    const openApiFactNodes = groupFacts({
      beforeFacts,
      afterFacts,
      changes: changelog,
    });

    const externalResults: RuleResult[] = [];
    for (const externalRule of externalRules) {
      const results = await externalRule.runRulesV2({
        ...inputs,
        groupedFacts: openApiFactNodes,
      });
      externalResults.push(...results);
    }

    // Run rules on specifications and collect the results
    const specificationResults: RuleResult[] = runSpecificationRules({
      specificationNode: openApiFactNodes.specification,
      rules: rulesOrRulesets,
      customRuleContext: context,
      beforeApiSpec,
      afterApiSpec,
    }).map(resultToRuleResult);

    const endpointResults: RuleResult[] = [];

    // For each endpoint from the endpoint fact nodes (this will include endpoints in both before and after specs) run rules and collect the results
    for (const endpointNode of openApiFactNodes.endpoints.values()) {
      const operationResults = runOperationRules({
        specificationNode: openApiFactNodes.specification,
        operationNode: endpointNode,
        rules: rulesOrRulesets,
        customRuleContext: context,
        beforeApiSpec,
        afterApiSpec,
      });
      endpointResults.push(...operationResults.map(resultToRuleResult));

      const requestRules = runRequestRules({
        specificationNode: openApiFactNodes.specification,
        operationNode: endpointNode,
        requestNode: endpointNode.request,
        rules: rulesOrRulesets,
        customRuleContext: context,
        beforeApiSpec,
        afterApiSpec,
      });
      endpointResults.push(...requestRules.map(resultToRuleResult));

      for (const responseNode of endpointNode.responses.values()) {
        const responseRules = runResponseRules({
          specificationNode: openApiFactNodes.specification,
          operationNode: endpointNode,
          responseNode: responseNode,
          rules: rulesOrRulesets,
          customRuleContext: context,
          beforeApiSpec,
          afterApiSpec,
        });

        const responseBodyRules = runResponseBodyRules({
          specificationNode: openApiFactNodes.specification,
          operationNode: endpointNode,
          responseNode: responseNode,
          rules: rulesOrRulesets,
          customRuleContext: context,
          beforeApiSpec,
          afterApiSpec,
        });
        endpointResults.push(
          ...responseBodyRules.map(resultToRuleResult),
          ...responseRules.map(resultToRuleResult)
        );
      }
    }

    return [...externalResults, ...specificationResults, ...endpointResults];
  }
}
