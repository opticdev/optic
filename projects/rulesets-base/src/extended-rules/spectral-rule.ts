import { jsonPointerHelpers } from '@useoptic/json-pointer-helpers';
import {
  IFact,
  IChange,
  OpenAPIV3,
  Result,
  constructFactTree,
  getFactForJsonPath,
} from '@useoptic/openapi-utilities';
import { ExternalRuleBase } from '../rules/external-rule-base';
import { isExempted } from '../rule-runner/utils';

type Lifecycle = 'added' | 'addedOrChanged' | 'changed' | 'always';

// Spectral results will be returned if they fail
export interface SpectralResult {
  path: (string | number)[];
  code: string | number;
  message: string;
}

export interface Spectral {
  run: (jsonSpec: any) => Promise<SpectralResult[]>;
}

function toOpticResult(
  spectralResult: SpectralResult,
  lifecycle: Lifecycle,
  relevantChangeOrFact: IFact | IChange,
  opts: {
    exempted: boolean;
    docsLink?: string;
  }
): Result {
  return {
    condition: spectralResult.code.toString(),
    exempted: opts.exempted,
    docsLink: opts.docsLink,
    passed: false,
    error: spectralResult.message,
    isMust: true,
    isShould: false,
    where: `${lifecycle} `,
    change: relevantChangeOrFact,
  };
}

export class SpectralRule extends ExternalRuleBase {
  private lifecycle: Lifecycle;
  private spectral: Spectral;
  public name: string;
  public docsLink?: string;
  constructor(options: {
    spectral: Spectral;
    name: string;
    applies?: Lifecycle;
    docsLink?: string;
  }) {
    super();
    this.name = options.name;
    this.spectral = options.spectral;
    this.lifecycle = options.applies ?? 'always';
    this.docsLink = options.docsLink;
  }

  async runRules(inputs: {
    context: any;
    nextFacts: IFact[];
    currentFacts: IFact[];
    changelog: IChange[];
    nextJsonLike: OpenAPIV3.Document<{}>;
    currentJsonLike: OpenAPIV3.Document<{}>;
  }): Promise<Result[]> {
    const factTree = constructFactTree(inputs.nextFacts);

    const changesByJsonPath: Record<string, IChange> = inputs.changelog.reduce(
      (acc, next) => {
        acc[next.location.jsonPath] = next;
        return acc;
      },
      {}
    );

    const spectralResults = await this.spectral.run(inputs.nextJsonLike);
    const results: Result[] = [];

    for (const spectralResult of spectralResults) {
      const path = jsonPointerHelpers.compile(
        spectralResult.path.map((p) => String(p))
      );
      const fact = getFactForJsonPath(path, factTree);
      if (!fact) {
        continue;
      }

      // This exemption is actually on the Fact level, rather than the spectral path
      // This is for consistency with our current rule engine. In the future we should attach exemptions on the nodes which trigger them, which would require us to rework the rules engine
      const rawForPath = jsonPointerHelpers.get(
        inputs.nextJsonLike,
        fact.location.jsonPath
      );
      const exempted = isExempted(rawForPath, this.name);

      // TODO in the future update to pass in the JSON path from spectral, rather than the fact json path
      if (this.lifecycle === 'always') {
        results.push(
          toOpticResult(spectralResult, 'always', fact, {
            exempted,
            docsLink: this.docsLink,
          })
        );
      } else {
        // find if there is an appropriate change
        const maybeChange: IChange | undefined =
          changesByJsonPath[fact.location.jsonPath];
        if (maybeChange) {
          if (this.lifecycle === 'added' && maybeChange.added) {
            results.push(
              toOpticResult(spectralResult, 'added', maybeChange, {
                exempted,
                docsLink: this.docsLink,
              })
            );
          } else if (this.lifecycle === 'changed' && maybeChange.changed) {
            results.push(
              toOpticResult(spectralResult, 'changed', maybeChange, {
                exempted,
                docsLink: this.docsLink,
              })
            );
          } else if (
            this.lifecycle === 'addedOrChanged' &&
            (maybeChange.added || maybeChange.changed)
          ) {
            results.push(
              toOpticResult(spectralResult, 'addedOrChanged', maybeChange, {
                exempted,
                docsLink: this.docsLink,
              })
            );
          }
        }
      }
    }
    return results;
  }
}
