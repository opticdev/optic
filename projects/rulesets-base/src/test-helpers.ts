import {
  OpenAPIV3,
  OpenAPITraverser,
  factsToChangelog,
  Result,
  defaultEmptySpec,
} from '@useoptic/openapi-utilities';
import { RuleRunner } from './rule-runner';
import { Rule, Ruleset } from './rules';
import { ExternalRuleBase } from './rules/external-rule-base';
import { OpenAPIDocument } from '.';

export const createRuleInputs = (
  beforeJson: OpenAPIDocument,
  afterJson: OpenAPIDocument
): Parameters<RuleRunner['runRulesWithFacts']>[0] => {
  const currentTraverser = new OpenAPITraverser();
  const nextTraverser = new OpenAPITraverser();

  currentTraverser.traverse(beforeJson);
  const currentFacts = [...currentTraverser.facts()];
  nextTraverser.traverse(afterJson);

  const nextFacts = [...nextTraverser.facts()];
  const changes = factsToChangelog(currentFacts, nextFacts);
  return {
    context: {},
    currentFacts,
    nextFacts,
    changelog: changes,
    currentJsonLike: beforeJson,
    nextJsonLike: afterJson,
  };
};

export const runRulesWithInputs = (
  rules: (Rule | Ruleset)[],
  beforeJson: OpenAPIDocument,
  afterJson: OpenAPIDocument
): Promise<Result[]> => {
  const ruleRunner = new RuleRunner(rules);
  return ruleRunner.runRulesWithFacts(createRuleInputs(beforeJson, afterJson));
};

export const externalRulesWithInputs = (
  rules: ExternalRuleBase,
  beforeJson: OpenAPIDocument,
  afterJson: OpenAPIDocument
): Promise<Result[]> => {
  const ruleRunner = new RuleRunner([rules]);
  return ruleRunner.runRulesWithFacts(createRuleInputs(beforeJson, afterJson));
};

export const createEmptySpec = (): OpenAPIV3.Document =>
  ({
    ...defaultEmptySpec,
    paths: {},
    info: {
      ...defaultEmptySpec.info,
    },
  }) as OpenAPIV3.Document;
