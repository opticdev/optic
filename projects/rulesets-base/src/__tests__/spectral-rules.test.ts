import { jest, test, expect, describe } from '@jest/globals'
import {
  Spectral,
  RulesetDefinition as SpectralRulesetDefinition,
} from '@stoplight/spectral-core';
import { oas } from '@stoplight/spectral-rulesets';
import { defaultEmptySpec, OpenAPIV3 } from '@useoptic/openapi-utilities';

import { SpectralRule } from '../extended-rules/spectral-rule';
import { createRuleInputs } from '../test-helpers';
import { RuleRunner } from '../rule-runner';

const emptySpec = {
  ...defaultEmptySpec,
  'x-optic-ci-empty-spec': true,
};

const specWithLintIssues: OpenAPIV3.Document = {
  ...defaultEmptySpec,
  servers: [{ url: 'http://optic.com' }],
  paths: {
    '/api/users': {
      get: {
        description: 'hello',
        responses: {},
      },
    },
  },
};

const specWithExemptions: OpenAPIV3.Document = {
  ...defaultEmptySpec,
  'x-optic-exemptions': 'spectral',
  servers: [{ url: 'http://optic.com' }],
  paths: {
    '/api/users': {
      get: {
        description: 'hello',
        responses: {},
      },
    },
  },
} as OpenAPIV3.Document;

describe('SpectralRule', () => {
  test('spectral rules run', async () => {
    const spectral = new Spectral();
    spectral.setRuleset(oas as SpectralRulesetDefinition);

    const ruleRunner = new RuleRunner([
      new SpectralRule({ spectral, name: 'spectral' }),
    ]);

    const results = await ruleRunner.runRulesWithFacts(
      createRuleInputs(specWithLintIssues, specWithLintIssues)
    );
    expect(results.every((result) => result.passed)).toBe(false);
    expect(results).toMatchSnapshot();
  });

  test('lifecycle rules can be applied to spectral rules', async () => {
    const spectral = new Spectral();
    spectral.setRuleset(oas as SpectralRulesetDefinition);

    const ruleRunner = new RuleRunner([
      new SpectralRule({ spectral, name: 'spectral', applies: 'added' }),
    ]);
    const resultsAdded = await ruleRunner.runRulesWithFacts(
      createRuleInputs(emptySpec, specWithLintIssues)
    );
    const resultsSame = await ruleRunner.runRulesWithFacts(
      createRuleInputs(specWithLintIssues, specWithLintIssues)
    );
    expect(resultsAdded.every((result) => result.passed)).toBe(false);
    expect(resultsSame.every((result) => result.passed)).toBe(true);
    expect(resultsAdded).toMatchSnapshot();
    expect(resultsSame).toMatchSnapshot();
  });

  test('can be exempted', async () => {
    const spectral = new Spectral();
    spectral.setRuleset(oas as SpectralRulesetDefinition);

    const ruleRunner = new RuleRunner([
      new SpectralRule({ spectral, name: 'spectral' }),
    ]);
    const results = await ruleRunner.runRulesWithFacts(
      createRuleInputs(specWithExemptions, specWithExemptions)
    );

    expect(results.every((result) => result.passed)).toBe(false);
    expect(results.some((result) => result.exempted)).toBe(true);
  });
});
