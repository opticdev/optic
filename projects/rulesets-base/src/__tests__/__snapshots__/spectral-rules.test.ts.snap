// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpectralRule lifecycle rules can be applied to spectral rules 1`] = `
Array [
  Object {
    "change": Object {
      "added": Object {
        "description": "hello",
        "method": "get",
        "pathPattern": "/api/users",
      },
      "changeType": "added",
      "location": Object {
        "conceptualLocation": Object {
          "method": "get",
          "path": "/api/users",
        },
        "conceptualPath": Array [
          "operations",
          "/api/users",
          "get",
        ],
        "jsonPath": "/paths/~1api~1users/get",
        "kind": "operation",
      },
    },
    "condition": "operation-operationId",
    "docsLink": undefined,
    "error": "Operation must have \\"operationId\\".",
    "exempted": false,
    "isMust": true,
    "isShould": false,
    "passed": false,
    "where": "added ",
  },
  Object {
    "change": Object {
      "added": Object {
        "description": "hello",
        "method": "get",
        "pathPattern": "/api/users",
      },
      "changeType": "added",
      "location": Object {
        "conceptualLocation": Object {
          "method": "get",
          "path": "/api/users",
        },
        "conceptualPath": Array [
          "operations",
          "/api/users",
          "get",
        ],
        "jsonPath": "/paths/~1api~1users/get",
        "kind": "operation",
      },
    },
    "condition": "operation-tags",
    "docsLink": undefined,
    "error": "Operation must have non-empty \\"tags\\" array.",
    "exempted": false,
    "isMust": true,
    "isShould": false,
    "passed": false,
    "where": "added ",
  },
  Object {
    "change": Object {
      "added": Object {
        "description": "hello",
        "method": "get",
        "pathPattern": "/api/users",
      },
      "changeType": "added",
      "location": Object {
        "conceptualLocation": Object {
          "method": "get",
          "path": "/api/users",
        },
        "conceptualPath": Array [
          "operations",
          "/api/users",
          "get",
        ],
        "jsonPath": "/paths/~1api~1users/get",
        "kind": "operation",
      },
    },
    "condition": "operation-success-response",
    "docsLink": undefined,
    "error": "Operation must have at least one \\"2xx\\" or \\"3xx\\" response.",
    "exempted": false,
    "isMust": true,
    "isShould": false,
    "passed": false,
    "where": "added ",
  },
]
`;

exports[`SpectralRule lifecycle rules can be applied to spectral rules 2`] = `Array []`;

exports[`SpectralRule spectral rules run 1`] = `
Array [
  Object {
    "change": Object {
      "location": Object {
        "conceptualLocation": Object {
          "method": "get",
          "path": "/api/users",
        },
        "conceptualPath": Array [
          "operations",
          "/api/users",
          "get",
        ],
        "jsonPath": "/paths/~1api~1users/get",
        "kind": "operation",
      },
      "value": Object {
        "description": "hello",
        "method": "get",
        "pathPattern": "/api/users",
      },
    },
    "condition": "operation-operationId",
    "docsLink": undefined,
    "error": "Operation must have \\"operationId\\".",
    "exempted": false,
    "isMust": true,
    "isShould": false,
    "passed": false,
    "where": "always ",
  },
  Object {
    "change": Object {
      "location": Object {
        "conceptualLocation": Object {
          "method": "get",
          "path": "/api/users",
        },
        "conceptualPath": Array [
          "operations",
          "/api/users",
          "get",
        ],
        "jsonPath": "/paths/~1api~1users/get",
        "kind": "operation",
      },
      "value": Object {
        "description": "hello",
        "method": "get",
        "pathPattern": "/api/users",
      },
    },
    "condition": "operation-tags",
    "docsLink": undefined,
    "error": "Operation must have non-empty \\"tags\\" array.",
    "exempted": false,
    "isMust": true,
    "isShould": false,
    "passed": false,
    "where": "always ",
  },
  Object {
    "change": Object {
      "location": Object {
        "conceptualLocation": Object {
          "method": "get",
          "path": "/api/users",
        },
        "conceptualPath": Array [
          "operations",
          "/api/users",
          "get",
        ],
        "jsonPath": "/paths/~1api~1users/get",
        "kind": "operation",
      },
      "value": Object {
        "description": "hello",
        "method": "get",
        "pathPattern": "/api/users",
      },
    },
    "condition": "operation-success-response",
    "docsLink": undefined,
    "error": "Operation must have at least one \\"2xx\\" or \\"3xx\\" response.",
    "exempted": false,
    "isMust": true,
    "isShould": false,
    "passed": false,
    "where": "always ",
  },
  Object {
    "change": Object {
      "location": Object {
        "conceptualLocation": Object {},
        "conceptualPath": Array [],
        "jsonPath": "",
        "kind": "specification",
      },
      "value": Object {
        "info": Object {
          "title": "Empty",
          "version": "0.0.0",
        },
        "openapi": "3.1.0",
        "servers": Array [
          Object {
            "url": "http://optic.com",
          },
        ],
      },
    },
    "condition": "info-contact",
    "docsLink": undefined,
    "error": "Info object must have \\"contact\\" object.",
    "exempted": false,
    "isMust": true,
    "isShould": false,
    "passed": false,
    "where": "always ",
  },
  Object {
    "change": Object {
      "location": Object {
        "conceptualLocation": Object {},
        "conceptualPath": Array [],
        "jsonPath": "",
        "kind": "specification",
      },
      "value": Object {
        "info": Object {
          "title": "Empty",
          "version": "0.0.0",
        },
        "openapi": "3.1.0",
        "servers": Array [
          Object {
            "url": "http://optic.com",
          },
        ],
      },
    },
    "condition": "info-description",
    "docsLink": undefined,
    "error": "Info \\"description\\" must be present and non-empty string.",
    "exempted": false,
    "isMust": true,
    "isShould": false,
    "passed": false,
    "where": "always ",
  },
]
`;
